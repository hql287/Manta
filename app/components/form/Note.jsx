// Libraries
import React, { Component } from 'react';
import PropTypes from 'prop-types';

// Custom Components
import { Section } from '../shared/Section';

// Animation
import _withFadeInAnimation from '../shared/hoc/_withFadeInAnimation';

// Styles
import styled from 'styled-components';
const NoteContent = styled.textarea`
  min-height: 36px;
  border-radius: 4px;
  padding: 10px;
  display: block;
  width: 100%;
  border: 1px solid #f2f3f4;
  color: #3a3e42;
  font-size: 14px;
`;

// Component
export class Note extends Component {
  constructor(props) {
    super(props);
    this.state = { content: this.props.note.content };
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.note.content === undefined) {
      this.setState({ content: '' }, () => {
        this.props.updateFieldData('note', this.state);
      });
    }
  }

  shouldComponentUpdate(nextProps, nextState) {
    if (this.state !== nextState) return true;
    if (this.props.note.content !== nextProps.note.content) return true;
    return false;
  }

  handleInputChange = (event) => {
    this.setState({ content: event.target.value }, () => {
      this.props.updateFieldData('note', this.state);
    });
  }

  render() {
    return (
      <Section>
        <label className="itemLabel">Note</label>
        <NoteContent
          cols="50"
          rows="4"
          onChange={this.handleInputChange}
          placeholder="Note"
          value={this.state.content}
        />
      </Section>
    );
  }
}

Note.propTypes = {
  note: PropTypes.object.isRequired,
  updateFieldData: PropTypes.func.isRequired,
};

// Export
export default _withFadeInAnimation(Note);
