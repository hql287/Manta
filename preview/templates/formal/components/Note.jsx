// React Libraries
import React from 'react';
import PropTypes from 'prop-types';

import styled from 'styled-components';

const InvoiceNote = styled.div `
  grid-column: 1 / 9;
  grid-row: 4;
  h1 {
    border-bottom: 2px solid #212529;
  }
  ${props => props.customAccentColor && `
    h1 { border-bottom: 2px solid ${props.accentColor};}
    `};

  h2 {
    padding-top: 1em;
  }

`;

// Component
function Note({t, invoice, configs}) {
  const {language, accentColor, customAccentColor} = configs;
  return invoice.note
    ? (
      <InvoiceNote accentColor={accentColor} customAccentColor={customAccentColor}>
        <h1>{t('preview:common:notice', {lng: language})}</h1>
        <h2>{invoice.note}</h2>
      </InvoiceNote>)
    : null;
}

Note.propTypes = {
  configs: PropTypes.object.isRequired,
  invoice: PropTypes.object.isRequired,
  t: PropTypes.func.isRequired
};

export default Note;
