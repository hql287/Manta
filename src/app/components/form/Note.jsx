// Libraries
import React, { Component } from 'react';
import PropTypes from 'prop-types';

// Custom Components
import { Section } from '../shared/Section';

// Animation
import _withFadeInAnimation from '../shared/hoc/_withFadeInAnimation';

// Styles
import styled from 'styled-components';
const NoteContent = styled.textarea`
  min-height: 36px;
  border-radius: 4px;
  padding: 10px;
  display: block;
  width: 100%;
  border: 1px solid #f2f3f4;
  color: #3a3e42;
  font-size: 14px;
`;

// Component
export class Note extends Component {
  constructor(props) {
    super(props);
    this.state = { content: this.props.note.content };
    this.handleInputChange = this.handleInputChange.bind(this);
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.note.content === undefined) {
      this.setState({ content: '' }, () => {
        this.props.updateFieldData('note', this.state);
      });
    }
  }

  shouldComponentUpdate(nextProps, nextState) {
    if (this.state !== nextState) return true;
    if (this.props.note.content !== nextProps.note.content) return true;
    return false;
  }

  handleInputChange(event) {
    this.setState({ content: event.target.value }, () => {
      this.props.updateFieldData('note', this.state);
    });
  }

  render() {
    const { t } = this.props;
    return (
      <Section>
        <label className="itemLabel">{t('form:fields:note')}</label>
        <NoteContent
          cols="50"
          rows="4"
          onChange={this.handleInputChange}
          value={this.state.content}
          placeholder={t('form:fields:note')}
        />
      </Section>
    );
  }
}

Note.propTypes = {
  note: PropTypes.object.isRequired,
  t: PropTypes.func.isRequired,
  updateFieldData: PropTypes.func.isRequired,
};

// Export
export default _withFadeInAnimation(Note);
