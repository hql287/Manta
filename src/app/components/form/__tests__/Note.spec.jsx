// Libs
import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';

// Component
import { Note } from '../Note.jsx';

// Mocks
const t = jest.fn();
const updateFieldData = jest.fn();
const note = {
  required: true,
  content: 'something',
};

describe('Note component', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = mount(<Note t={t} note={note} updateFieldData={updateFieldData} />);
  });

  // PROPS & STATE
  it('receives correct props', () => {
    expect(wrapper.prop('note')).toEqual(note);
    expect(wrapper.prop('updateFieldData')).toEqual(updateFieldData);
  });

  // RENDER
  it('renders necessary element', () => {
    expect(wrapper.find('label')).toHaveLength(1);
    expect(wrapper.find('textarea')).toHaveLength(1);
  });

  // LIFE CYCLE EVENTS
  // TODO
  it('render when necessary', () => {});

  // PRIVATE METHOD
  it('handleInputChange correctly', () => {
    const spy = jest.spyOn(Note.prototype, 'handleInputChange');
    const wrap = mount(<Note t={t} note={note} updateFieldData={updateFieldData} />);
    const textArea = wrap.find('textarea');
    textArea.simulate('change', { target: { value: 'Something Else' } });
    expect(spy).toHaveBeenCalled();
  });

  // SNAPSHOT
  it('matches snapshot', () => {
    const tree = renderer
      .create(<Note t={t} note={note} updateFieldData={updateFieldData} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
