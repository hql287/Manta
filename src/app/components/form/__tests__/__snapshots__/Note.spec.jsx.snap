// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Note component matches snapshot 1`] = `
<div
  className="Section__SectionStyle-dgSKSv cwfAAu"
>
  <label
    className="itemLabel"
  />
  <textarea
    className="Note__NoteContent-enQLsi haIqin"
    cols="50"
    onChange={[Function]}
    placeholder={undefined}
    rows="4"
    value="something"
  />
</div>
`;
